dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/cloudera/manager/AbstractClouderaYarnApplicationTask.java

```diff
@@ -0,0 +1,131 @@
+ */
+package com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager;
+
+import com.google.common.base.Preconditions;
+import com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.dto.ApiYARNApplicationDTO;
+import com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.dto.ApiYARNApplicationListDTO;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.impl.client.CloseableHttpClient;
+
+public abstract class AbstractClouderaYarnApplicationTask extends AbstractClouderaManagerTask {
+  private static final DateTimeFormatter isoDateTimeFormatter =
+      DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
+
+  private final String fromDate;
+
+  public AbstractClouderaYarnApplicationTask(String fileName, int includedLastDays) {
+    super(String.format("%s-%dd.jsonl", fileName, includedLastDays));
+    Preconditions.checkArgument(
+        includedLastDays > 1,
+        String.format("Amount of days must be a positive number. Get %d.", includedLastDays));
+    this.fromDate = buildISODateTime(includedLastDays);
+  }
+
+  private String buildISODateTime(int deltaInDays) {
+    ZonedDateTime dateTime =
+        ZonedDateTime.of(LocalDateTime.now().minusDays(deltaInDays), ZoneId.of("UTC"));
+    return dateTime.format(isoDateTimeFormatter);
+  }
+
+  class PaginatedClouderaYarnApplicationsLoader {
+    private final String host;
+    private final CloseableHttpClient httpClient;
+    private final int limit;
+    private int offset;
+
+    public PaginatedClouderaYarnApplicationsLoader(ClouderaManagerHandle handle) {
+      this.host = handle.getApiURI().toString();
+      this.httpClient = handle.getHttpClient();
+      this.limit = 500;
+      this.offset = 0;
+    }
+
+    public int load(String clusterName, Consumer<List<ApiYARNApplicationDTO>> callback) {
+      return load(clusterName, null, callback);
+    }
+
+    public int load(
+        String clusterName,
+        @Nullable String appType,
+        Consumer<List<ApiYARNApplicationDTO>> callback) {
+      int amountOfLoadedApps = 0;
+      offset = 0;
+      boolean nextLoad = true;
+
+      while (nextLoad) {
+        URI yarnAppsURI = buildYARNApplicationURI(clusterName, appType);
+        List<ApiYARNApplicationDTO> newLoad = load(yarnAppsURI);
+        if (!newLoad.isEmpty()) {
+          callback.accept(newLoad);
+        } else {
+          nextLoad = false;
+        }
+        amountOfLoadedApps += newLoad.size();
+        offset += limit;
```

Comment:
Should it be `newLoad.size()` instead of `limit`?

Reply:
Actually it doesn't matter. It's expected that API returns no more than `limit` rows per request. And in the last call the API should return an empty list.